### PR TITLE
Refactor pixel snapping.

### LINKED ISSUE #41814: Godot Screen Pixel Jitter.
**Godot version:**
Godot 3.2.2 stable, and Godot 4.0
GLES 2,3 Vulkan

**OS/device including version:**

Windows 10 1909 , GTX 1070, I5 , 1360x780 screen.

**Issue description:**

Pixels snap randomly and offest and scaling isn't smooth and jitter, also screen differnets can snap the pixels randomly.
as8
as9

Pixel snap has no effect at all.
nor making the values smaller in godot source code.

**Steps to reproduce:**
Simply try to scale a sprite very slowly and you see line of tearing visible.
or change the viewport size while looking at your sprites.

**Minimal reproduction project:**
[jitter.zip](https://github.com/godotengine/godot/files/5179932/jitter.zip)

### PR DESCRIPTION
**Pull request by reduz**
-Rename pixel_snap to snap_2d_to_vertices
-Added snap_2d_to_transforms which is more useful

Fixes #41814
Solves proposal https://github.com/godotengine/godot-proposals/issues/1666
Supersedes #35606, supersedes #41535, supersedes #41534

### CODE REVIEW

**Comment by RamaStudio** [comment]
I have done testing this Pull Request.
Using :
Godot 3.2.3,
Godot 4.0calinou.6c173e2f7 (Aka older version of Godot 4.0) ,
Godot 4.0 Pixel snap refactor


Here is what I came up with :


First I notices this Shaking effect in the new pixel snap refactor, its very noticeable.
I thinks its due to the snapping that happens in Realtime.

Godot 4.0 Refactor :


New Godot 1


Old Godot 4,0:

Old Godot 1

Godot 3.2.3 : 

Godot 1

Also the smoothness of movement in the Refactor Godot is very blocky.
Old pixel snap didn't have this "blackness" as shown in the GIFs


note: its stated by Calinou in #35635 That by snapping the camera you lose smoothness of motion.
Could be the reason, and I don't think there is a workaround for it.


Also when doing my tests I find that Godot 3.2.3 and Godot 4.0 are different in how they render the sprites.

Differnce


Could be because of changes in Vulkan and HDR but that's not our talk, i just stated it because it give me a problem comparing the changes.(and also that I think Godot 3.2.3 sprites looks sharper then Godot 4.0)

Another problem I find is a wave like seam that appears in the sprite while moving,
I couldn't able to capture it using Gifs, but if you stop just in the right time, your sprite will get distorted.
This problem happens in both Godot 4.0 and doesn't happen in Godot 3.2.3:

Screenshot 2020-10-30 003639

As you can see the eye of my sprite got extra pixel on the left eye.

the sprite get distorted: unlike Godot 3.2.3 the sprite remain the same no matter how you move.
(could be because the sprites are not getting snapped as a whole, but some pixels get?)
Same for all sprites.
So it doesn't really pixel snap anything but just shake up the camera or something.


What I think is this Refactor pixel snapping was based on Godot 4.0 pixel snapping which my tests show is broken.
So if you did Refactor this tool based on Godot 4.0 then its broken. and should be refactored based on Godot 3.2
as I see in the code, a lot of changes in the sprite code between Godot 3.2 and Godot 4.0.
So I can't really tell what did go wrong.

Hope this help address the problems of pixel snapping in Godot 4.0, and improves pixel snapping of Godot 3.2.

**Comment by reduz** [comment]
@RamaStudio I am not exactly sure what I should be looking at. 4.0 prior to this PR did not have pixel snap implemented, so even if you turn on the setting it was the same as just disabling it (made no difference).

If you are using filtered textures, you should probably not enable snapping. Snapping will be mostly useful for pixel art. This PR aims to solve the problems with motion (see the referenced issues/PRS0, not with scaling.

Pixel (no filter) scaling will always be wobbly, not sure if anything can be done about that.

**Comment by reduz** [comment]
Also you should most definitely not be using vertex to pixel snap for pixel art. Use the new transforms to pixel snap instead.

**Comment by RamaStudio** [comment]
> @RamaStudio I am not exactly sure what I should be looking at. 4.0 prior to this PR did not have pixel snap implemented, so even if you turn on the setting it was the same as just disabling it (made no difference).

That's explain why its broken.

> If you are using filtered textures, you should probably not enable snapping. Snapping will be mostly useful for pixel art. This PR aims to solve the problems with motion (see the referenced issues/PRS0, not with scaling.

I didn't. but some how in Godot 4.0 the sprites look more linear even with filtering set to nearest.
that could be why you thought I didn't use it.

> Pixel (no filter) scaling will always be wobbly, not sure if anything can be done about that.

As for the motion problem I already said that it has the problem of "not snapping anything, and causing wobbly"

Here is what I meant:
Godot 1


The the sprites changes every time I stops : 
Untitled


Now look Godot 3.2.3 :
Godot 2


The sprite never changes with pixel snap on. (It may get distorted at some scale conditions but its way fewer wobbly effect then 4.0)

that's the only bug I can report here.


>Pixel (no filter) scaling will always be wobbly, not sure if anything can be done about that.

I didn't say it was wobbly, I said it was shaking. compare it with the Gifs I sent. there is up down left right movement to the camera or to all the sprites.
maybe the pixels are getting snapped aggressively or something.

>Also you should most definitely not be using vertex to pixel snap for pixel art. Use the new transforms to pixel snap instead.

I checked your code to see what that option does, couldn't find anything. but I kept it disabled anyway.

**Comment by reduz** [comment]
Ok just so we are on the same page, there are two options here:
image

The first one is meant to solve #41535. With it, the demos run ok, no jitter between objects reaching the next pixel at different times. When you turn it on, all reach the next pixel at the same time. This is definitely meant for pixel art, if your texture has filtering enabled you may not want to use this.

The second is the same as the old "pixel snap" in Godot 3.x and should work the same. It wasn't being enabled but it should use the same code as 3.2 so I assume it works. If it does not, it's most likely a bug and it needs to be fixed.

**Comment by RPicster** [comment]
I would love to test this. It's a huge problem in my game.
Unfortunately I still did not understand how to make custom builds including certain PRs.
If someone has a Win build on 3.2 "lying around" with this change, I would love to test it and give feedback how it behaves with my project.

File: scene/main/viewport.cpp [hunk 1]
```diff
@@ -3419,6 +3440,8 @@ void Viewport::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "world_2d", PROPERTY_HINT_RESOURCE_TYPE, "World2D", 0), "set_world_2d", "get_world_2d");
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "transparent_bg"), "set_transparent_background", "has_transparent_background");
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "handle_input_locally"), "set_handle_input_locally", "is_handling_input_locally");
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "snap_2d_transforms_to_pixel"), "set_snap_2d_transforms_to_pixel", "is_snap_2d_transforms_to_pixel_enabled");
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "snap_2d_vertices_to_pixel"), "set_snap_2d_vertices_to_pixel", "is_snap_2d_vertices_to_pixel_enabled");
```
Comment (akien-mga) [inline scene/main/viewport.cpp:3444]: Seems like you forgot the method bindings for those?

File: scene/main/viewport.cpp [hunk 2]
```diff
@@ -3577,6 +3600,8 @@ Viewport::Viewport() {
 	debug_draw = DEBUG_DRAW_DISABLED;
 
 	snap_controls_to_pixels = true;
+	snap_2d_transforms_to_pixel = false;
```
Comment (akien-mga) [inline scene/main/viewport.cpp:3603]: You forgot to initialize the other one.

**Comment by Xrayez** [comment]
@RPicster builds artifacts are available for this PR to test:

Windows: https://github.com/godotengine/godot/actions/runs/336674241

Go into "Show all checks" → "Details" for each platform of interest → "Artifacts".

**Comment by Calinou** [comment]
@RPicster This pull request is based on the `master` branch (with the Vulkan renderer), not the `3.2` branch. It's not compatible with 3.2.x projects.

**Review by akien-mga — APPROVED** [review]

**Comment by RamaStudio** [comment]
@reduz 
With that idea in mine I guess everything works fine,
I did further tests with bigger viewport size. and it shows no pixel artifact at all. (the wobbly movement is removed, if you set the viewport size the same as the window size.)

However the shaking problem I talked about is confirmed.
and happens when you enable "snap 2d transformation to pixel"

and that's it. 👍 
Thank you.

**Comment by Zireael07** [comment]
Can the "viewport size = window size" requirement for pixel perfect get documented?

File: servers/rendering/rendering_server_canvas.cpp [hunk 3]
```diff
@@ -113,7 +113,12 @@ void RenderingServerCanvas::_cull_canvas_item(Item *p_canvas_item, const Transfo
 	}
 
 	Rect2 rect = ci->get_rect();
-	Transform2D xform = p_transform * ci->xform;
+	Transform2D xform = ci->xform;
+	if (snapping_2d_transforms_to_pixel) {
+		xform.elements[2].floor();
```
Comment (lawnjelly) [inline servers/rendering/rendering_server_canvas.cpp:118]: I've just been backporting this to 3.2, and noticed, there is an error here which stops the whole method from working.

This line needs to be `xform.elements[2] = xform.elements[2].floor();`, as it stands, this line is a no-op. I've made this mistake more than once with godot funcs myself! :grin:

**Comment by lawnjelly** [comment]
@reduz I've pointed out small typo which may be needed for this to work in master branch. I can't currently run 4.0 so anyone on 4.0 (@akien-mga ?) can make a small PR to fix this.

**Comment by aaronfranke** [comment]
@reduz Should those issues you linked be closed?

**Comment by lawnjelly** [comment]
> @reduz Should those issues/PRs you linked be closed?

For the issues at least an equivalent PR still has to be finalized / merged for 3.2 first I think (this one can't be cherry picked super easily because there are some differences).

**Pull request merged**